x/oracle/genesis.go

```diff
@@ -9,12 +11,107 @@ import (
 
 // InitGenesis initializes the x/oracle module's state from a provided genesis
 // state.
-func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
-	k.SetParams(ctx, genState.Params)
+func InitGenesis(ctx sdk.Context, keeper keeper.Keeper, genState types.GenesisState) {
+	for _, d := range genState.FeederDelegations {
+		voter, err := sdk.ValAddressFromBech32(d.ValidatorAddress)
+		if err != nil {
+			panic(err)
+		}
+
+		feeder, err := sdk.AccAddressFromBech32(d.FeederAddress)
+		if err != nil {
+			panic(err)
+		}
+
+		keeper.SetFeederDelegation(ctx, voter, feeder)
+	}
+
+	for _, ex := range genState.ExchangeRates {
+		keeper.SetExchangeRate(ctx, ex.Denom, ex.ExchangeRate)
+	}
+
+	for _, mc := range genState.MissCounters {
+		operator, err := sdk.ValAddressFromBech32(mc.ValidatorAddress)
+		if err != nil {
+			panic(err)
+		}
+
+		keeper.SetMissCounter(ctx, operator, mc.MissCounter)
+	}
+
+	for _, ap := range genState.AggregateExchangeRatePrevotes {
+		valAddr, err := sdk.ValAddressFromBech32(ap.Voter)
+		if err != nil {
+			panic(err)
+		}
+
+		keeper.SetAggregateExchangeRatePrevote(ctx, valAddr, ap)
+	}
+
+	for _, av := range genState.AggregateExchangeRateVotes {
+		valAddr, err := sdk.ValAddressFromBech32(av.Voter)
+		if err != nil {
+			panic(err)
+		}
+
+		keeper.SetAggregateExchangeRateVote(ctx, valAddr, av)
+	}
```

Comment:
Does this mean that genesis state contains initial prices, as a consequence of hard-coded votes from validator addresses?

Reply:
I think the idea is this happens on export/import network upgrades. This process is old and somewhat antiquated, in favor of live upgrades via `x/upgrade`. So TLDR; in practice this code will never run except in maybe some tests cases.

Reply:
It contains the initial exchange rate and votes for them, but the votes don't turn into exchange rates until after the ballot is counted -- the exchange rates will always be at least one vote period behind the votes